README.md
# Bencode

[![Build Status](https://travis-ci.org/avli/bencoder.svg?branch=master)](https://travis-ci.org/avli/bencoder)

A [Bencode](https://en.wikipedia.org/wiki/Bencode) encoding/decoding library written in TypeScript.

# Usage

```javascript
Bencoder = require('bencoder')

// Encode data

var x1 = Bencoder.encode("foobar")
var x2 = Bencoder.encode(42)
var x3 = Bencoder.encode({foo: [1, 2, 3], bar: 42, baz: "foobarbaz"})

// x1, x2, and x3 are Buffer objects. For example x1 is actually <Buffer 36 3a 66 6f 6f 62 61 72>
// To see the string representation do

x1.toString()
'6:foobar'

x2.toString()
'i42e'

x3.toString()
'd3:bari42e3:baz9:foobarbaz3:food1:0i1e1:1i2e1:2i3eee'

// Decode data

Bencoder.decode(x1)
'foobar'
Bencoder.decode(x2)
42
Bencoder.decode(x3)
{ bar: 42, baz: 'foobarbaz', foo: [ 1, 2, 3 ] }

// A few words about encodings
// By default bencoder assumes everything is UTF-8. For example

var x4 = Bencoder.encode('привет')
Bencoder.decode(x4) // which is the same as Bencoder.decode(x4, 'utf8')
'привет'

// Or you can specify another encoding, if you want to
Bencoder.decode(x4, 'ascii')
'P?Q\u0000P8P2P5Q\u0002' 

```



CHANGELOG.md
# Version 0.0.5

Add typings.

# Version 0.0.4

Add encodings support.


test/test.js
var assert = require('assert');
var bencoder = require('../out/bencoder')

describe('bencoder', function() {
    describe('#encode()', function() {
        it('should encode "foo" to "3:foo"', function () {
            assert.equal("3:foo", bencoder.encode('foo').toString());
        });
        it('should encode an empty string to "0:"', function () {
            assert.equal("0:", bencoder.encode('').toString());
        });
        it('should encode 42 to "i42e"', function () {
            assert.equal("i42e", bencoder.encode(42).toString());
        });
    });

    describe('#decode()', function() {
        it('should dencode "3:foo" to "foo"', function () {
            assert.equal('foo', bencoder.decode(Buffer.from('3:foo')).toString());
        });
        it('should raise exception when decoding "3:foobar"', function () {
            assert.throws(() => {
                bencoder.decode(Buffer.from('3:foobar')).toString()
            }, 
            Error);
        });
    });
})



test/test.ts
'use strict';

import * as assert from 'assert';
import { Buffer } from 'buffer';
import * as bencoder from '../bencoder'

describe('bencoder', () => {
    describe('#encode()', () => {
        it('should encode "foo" to "3:foo"', () => {
            assert.equal("3:foo", bencoder.encode('foo').toString());
        });
        it('should encode an empty string to "0:"', () => {
            assert.equal("0:", bencoder.encode('').toString());
        });
        it('should encode 42 to "i42e"', () => {
            assert.equal("i42e", bencoder.encode(42).toString());
        });
        it('should encode [1, 2, 3] to "li1ei2ei3ee"', () => {
            assert.equal("li1ei2ei3ee", bencoder.encode([1, 2, 3]).toString());
        });
        it('should encode [1, "2", "foo", 42] to li1e1:23:fooi42ee', () => {
            assert.equal("li1e1:23:fooi42ee", bencoder.encode([1, "2", "foo", 42]).toString());
        });
        it('should encode {"a": 1, "b": 2, 3: 42} to d1:3i42e1:ai1e1:bi2ee', () => {
            assert.equal("d1:3i42e1:ai1e1:bi2ee", bencoder.encode({ "a": 1, "b": 2, 3: 42 }).toString());
        });
        it('should encode {"list": [1, 2, 3]} to "d4:listli1ei2ei3eee"', () => {
            assert.equal("d4:listli1ei2ei3eee", bencoder.encode({ "list": [1, 2, 3] }).toString());
        });
        it('should encode [{d1: {a: 1}}, {d2: {b: 2}}] to "ld2:d1d1:ai1eed2:d2d1:bi2eeee"', () => {
            assert.equal("ld2:d1d1:ai1eeed2:d2d1:bi2eeee", bencoder.encode([{ d1: { a: 1 } }, { d2: { b: 2 } }]).toString());
        });
        it('should encode {foo: [1, 2, 3], bar: 42, baz: "foobarbaz"} to "d3:bari42e3:baz9:foobarbaz3:fooli1ei2ei3eee"', () => {
            assert.deepEqual("d3:bari42e3:baz9:foobarbaz3:fooli1ei2ei3eee", bencoder.encode({ foo: [1, 2, 3], bar: 42, baz: "foobarbaz" }).toString());
        });
    });

    describe('#decode()', () => {
        it('should dencode "3:foo" to "foo"', () => {
            assert.equal('foo', bencoder.decode(Buffer.from('3:foo')).toString());
        });
        it('should decode "li1ei2e5:threee" to [1, 2, "three"]', () => {
            assert.deepEqual([1, 2, "three"], bencoder.decode(Buffer.from("li1ei2e5:threee")));
        });
        it('should raise exception when decoding "li42eee"', () => {
            assert.throws(() => {
                bencoder.decode(Buffer.from("li42eee"))
            },
                (err) => {
                    return err.message === 'Unexpected continuation: "e"'
                });
        });
        it('should raise exception when decoding "li42e"', () => {
            assert.throws(() => {
                bencoder.decode(Buffer.from("li42e"))
            });
        });
        it('should decode "d1:ai1e3:fooi42e" to {a: 1, foo: 42}', () => {
            assert.deepEqual({ a: 1, 'foo': 42 }, bencoder.decode(Buffer.from("d1:ai1e3:fooi42e")))
        });
        it('should decode "d3:food3:bar3:bazee" to {foo: {bar: "baz"}}', () => {
            assert.deepEqual({ foo: { bar: "baz" } }, bencoder.decode(Buffer.from("d3:food3:bar3:bazee")))
        })
        it('it should raise exception when decoding "d3:food3bar3:bazee"', () => {
            assert.throws(() => {
                bencoder.decode(Buffer.from("d3:food3bar3:bazee"))
            },
                (err) => {
                    assert.equal(err.message, "Expected d, i, l or digit, got \"ee\"");
                    return true;
                });
        });
    });

    describe('#decodeString()', () => {
        it('should raise exception when decoding "3foo"', () => {
            assert.throws(() => {
                bencoder.decode(Buffer.from('3foo'))
            },
                Error);
        });
        it('should return "привет" when decoding "12:привет"', () => {
            assert.equal("привет", bencoder.decode(Buffer.from("12:привет"), 'utf8'));
        })
    })

    describe('#decodeInteger()', () => {
        it('should raise exception when decoding "i42"', () => {
            assert.throws(() => {
                bencoder.decode(Buffer.from('i42'))
            },
                Error);
        });
    })
})


out/test/test.js
'use strict';
var assert = require('assert');
var buffer_1 = require('buffer');
var bencoder = require('../bencoder');
describe('bencoder', function () {
    describe('#encode()', function () {
        it('should encode "foo" to "3:foo"', function () {
            assert.equal("3:foo", bencoder.encode('foo').toString());
        });
        it('should encode an empty string to "0:"', function () {
            assert.equal("0:", bencoder.encode('').toString());
        });
        it('should encode 42 to "i42e"', function () {
            assert.equal("i42e", bencoder.encode(42).toString());
        });
        it('should encode [1, 2, 3] to "li1ei2ei3ee"', function () {
            assert.equal("li1ei2ei3ee", bencoder.encode([1, 2, 3]).toString());
        });
        it('should encode [1, "2", "foo", 42] to li1e1:23:fooi42ee', function () {
            assert.equal("li1e1:23:fooi42ee", bencoder.encode([1, "2", "foo", 42]).toString());
        });
        it('should encode {"a": 1, "b": 2, 3: 42} to d1:3i42e1:ai1e1:bi2ee', function () {
            assert.equal("d1:3i42e1:ai1e1:bi2ee", bencoder.encode({ "a": 1, "b": 2, 3: 42 }).toString());
        });
        it('should encode {"list": [1, 2, 3]} to "d4:listli1ei2ei3eee"', function () {
            assert.equal("d4:listli1ei2ei3eee", bencoder.encode({ "list": [1, 2, 3] }).toString());
        });
        it('should encode [{d1: {a: 1}}, {d2: {b: 2}}] to "ld2:d1d1:ai1eed2:d2d1:bi2eeee"', function () {
            assert.equal("ld2:d1d1:ai1eeed2:d2d1:bi2eeee", bencoder.encode([{ d1: { a: 1 } }, { d2: { b: 2 } }]).toString());
        });
        it('should encode {foo: [1, 2, 3], bar: 42, baz: "foobarbaz"} to "d3:bari42e3:baz9:foobarbaz3:fooli1ei2ei3eee"', function () {
            assert.deepEqual("d3:bari42e3:baz9:foobarbaz3:fooli1ei2ei3eee", bencoder.encode({ foo: [1, 2, 3], bar: 42, baz: "foobarbaz" }).toString());
        });
    });
    describe('#decode()', function () {
        it('should dencode "3:foo" to "foo"', function () {
            assert.equal('foo', bencoder.decode(buffer_1.Buffer.from('3:foo')).toString());
        });
        it('should decode "li1ei2e5:threee" to [1, 2, "three"]', function () {
            assert.deepEqual([1, 2, "three"], bencoder.decode(buffer_1.Buffer.from("li1ei2e5:threee")));
        });
        it('should raise exception when decoding "li42eee"', function () {
            assert.throws(function () {
                bencoder.decode(buffer_1.Buffer.from("li42eee"));
            }, function (err) {
                return err.message === 'Unexpected continuation: "e"';
            });
        });
        it('should raise exception when decoding "li42e"', function () {
            assert.throws(function () {
                bencoder.decode(buffer_1.Buffer.from("li42e"));
            });
        });
        it('should decode "d1:ai1e3:fooi42e" to {a: 1, foo: 42}', function () {
            assert.deepEqual({ a: 1, 'foo': 42 }, bencoder.decode(buffer_1.Buffer.from("d1:ai1e3:fooi42e")));
        });
        it('should decode "d3:food3:bar3:bazee" to {foo: {bar: "baz"}}', function () {
            assert.deepEqual({ foo: { bar: "baz" } }, bencoder.decode(buffer_1.Buffer.from("d3:food3:bar3:bazee")));
        });
        it('it should raise exception when decoding "d3:food3bar3:bazee"', function () {
            assert.throws(function () {
                bencoder.decode(buffer_1.Buffer.from("d3:food3bar3:bazee"));
            }, function (err) {
                assert.equal(err.message, "Expected d, i, l or digit, got \"ee\"");
                return true;
            });
        });
    });
    describe('#decodeString()', function () {
        it('should raise exception when decoding "3foo"', function () {
            assert.throws(function () {
                bencoder.decode(buffer_1.Buffer.from('3foo'));
            }, Error);
        });
        it('should return "привет" when decoding "12:привет"', function () {
            assert.equal("привет", bencoder.decode(buffer_1.Buffer.from("12:привет"), 'utf8'));
        });
    });
    describe('#decodeInteger()', function () {
        it('should raise exception when decoding "i42"', function () {
            assert.throws(function () {
                bencoder.decode(buffer_1.Buffer.from('i42'));
            }, Error);
        });
    });
});
//# sourceMappingURL=test.js.map

bencoder.ts
'use strict';

import { Buffer } from 'buffer';

/*
Encoding
*/

function encodeString(s: string): Buffer {
    let bytes = Buffer.from(s);
    return Buffer.concat([
        Buffer.from(bytes.length.toString()),
        Buffer.from(':'),
        bytes]);
}

function encodeInteger(i: number): Buffer {
    return Buffer.from(`i${i}e`);
}

function encodeArray(l: Array<any>): Buffer {
    let result: Array<Buffer> = [Buffer.from('l')];
    l.forEach(element => {
        result.push(_encode(element));
    });
    result.push(Buffer.from('e'))
    return Buffer.concat(result);
}

function encodeDict(d: any): Buffer {
    let result: Array<Buffer> = [Buffer.from('d')];
    let keys = Object.keys(d).sort();
    keys.forEach(k => {
        result.push(encodeString(k));
        result.push(encode(d[k]))
    });
    result.push(Buffer.from('e'));
    return Buffer.concat(result);
}

function _encode(data: any): Buffer {
    switch (typeof data) {
        case 'string':
            return encodeString(data);
        case 'number':
            return encodeInteger(data);
        case 'object':
            if (Array.isArray(data)) {
                return encodeArray(data);
            }
            else {
                return encodeDict(data);
            }
    }
}


/**
 * Encodes data in bencode.
 *
 * @param {Array|String|Object|Number} data
 * @return {Buffer}
 */
export function encode(data: any): Buffer {
    return new Buffer(_encode(data));
}

/*
Decoding
*/

enum Delimeters {
    i = 0x69,
    e = 0x65,
    l = 0x6c,
    d = 0x64,
    colon = 0x3a
}

interface DecodingResult {
    value: any,
    rest?: Buffer
}

function decodeString(data: Buffer, encoding?: string): DecodingResult {
    let firstColonIndex = null;
    for (let i = 0; i < data.length; i++) {
        if (data[i] === Delimeters.colon) {
            firstColonIndex = i;
            break;
        }
    }

    if (!firstColonIndex) {
        throw new Error(`Error while decoding ${data.toString(encoding)}. Missing colon.`);
    }

    let expectedLength = parseInt(data.slice(0, firstColonIndex).toString());
    let value = data.slice(firstColonIndex + 1, firstColonIndex + 1 + expectedLength);
    return { value: value.toString(encoding), rest: data.slice(firstColonIndex + 1 + expectedLength) }
}

function decodeInteger(data: Buffer): DecodingResult {
    let endIndex: number = null // The first 'e' symbol index
    for (let i = 1; i < data.length; i++) {
        if (data[i] === Delimeters.e) {
            endIndex = i;
            break;
        }
    }

    if (!endIndex) {
        throw new Error("Invalid data. 'e' symbol expected");
    }

    return {
        value: parseInt(data.slice(1, endIndex).toString()),
        rest: data.slice(endIndex + 1)
    }
}

function decodeList(data: Buffer, encoding?: string): DecodingResult {
    // TODO: Check for 'e' symbol in the end.
    let result = [];
    let rest = data.slice(1); // l...
    let value = null;

    while (true) {
        let firstByte = rest[0];
        if (firstByte === Delimeters.i) {
            ({ value, rest } = decodeInteger(rest));
            result.push(value);
            continue;
        }
        if (encodesDigit(firstByte)) {
            ({ value, rest } = decodeString(rest, encoding));
            result.push(value);
            continue;
        }
        if (firstByte === Delimeters.l) {
            ({ value, rest } = decodeList(rest, encoding));
            result.push(value);
            continue;
        }
        if (firstByte === Delimeters.d) {
            ({ value, rest } = decodeDict(rest, encoding));
            result.push(value);
            continue;
        }
        if (firstByte === Delimeters.e) { // end of the list
            rest = rest.slice(1);
            break;
        }
        throw new Error(`Expected d, i, l or digit, got ${rest.toString(encoding)}`)
    }
    return { value: result, rest: rest };
}

function decodeDict(data: Buffer, encoding?: string): DecodingResult {
    let result = {};
    let rest = data.slice(1); // d...
    let value = null;
    let key: string;

    while (rest.length !== 0) {
        let firstByte = rest[0];
        if (firstByte === Delimeters.e) {
            rest = rest.slice(1);
            break;
        }
        ({ value: key, rest } = decodeString(rest, encoding));
        ({ value, rest } = _decode(rest, encoding));
        result[key] = value;
    }

    return { value: result, rest: rest }
}

function encodesDigit(x: number) {
    return (x >= 0x30) && (x <= 0x39);
}

function _decode(data: Buffer, encoding?: string): any {
    let value = null;
    let rest = null;
    let firstByte = data[0];
    if (firstByte === Delimeters.i) {
        return decodeInteger(data);
    }
    else if (encodesDigit(firstByte)) {
        return decodeString(data, encoding);
    }
    else if (firstByte === Delimeters.l) {
        return decodeList(data, encoding);
    }
    else if (firstByte === Delimeters.d) {
        return decodeDict(data, encoding);
    }
    else {
        throw new Error(`Expected d, i, l or digit, got "${data.toString()}"`);
    }
}

/**
 * Decodes bencoded data.
 *
 * @param {Buffer} data
 * @param {String} encoding (optional)
 * @return {Object|Array|String|Number}
 */
export function decode(data: Buffer, encoding?: string): any {
    let value: string;
    let rest: Buffer
    ({ value, rest } = _decode(data, encoding));
    if (rest.length) {
        throw new Error(`Unexpected continuation: "${rest.toString()}"`)
    }
    return value;
}


bencoder.js
'use strict';
var buffer_1 = require('buffer');
/*
Encoding
*/
function encodeString(s) {
    var bytes = buffer_1.Buffer.from(s);
    return buffer_1.Buffer.concat([
        buffer_1.Buffer.from(bytes.length.toString()),
        buffer_1.Buffer.from(':'),
        bytes]);
}
function encodeInteger(i) {
    return buffer_1.Buffer.from("i" + i + "e");
}
function encodeArray(l) {
    var result = [buffer_1.Buffer.from('l')];
    l.forEach(function (element) {
        result.push(_encode(element));
    });
    result.push(buffer_1.Buffer.from('e'));
    return buffer_1.Buffer.concat(result);
}
function encodeDict(d) {
    var result = [buffer_1.Buffer.from('d')];
    var keys = Object.keys(d).sort();
    keys.forEach(function (k) {
        result.push(encodeString(k));
        result.push(encode(d[k]));
    });
    result.push(buffer_1.Buffer.from('e'));
    return buffer_1.Buffer.concat(result);
}
function _encode(data) {
    switch (typeof data) {
        case 'string':
            return encodeString(data);
        case 'number':
            return encodeInteger(data);
        case 'object':
            if (Array.isArray(data)) {
                return encodeArray(data);
            }
            else {
                return encodeDict(data);
            }
    }
}
/**
 * Encodes data in bencode.
 *
 * @param {Array|String|Object|Number} data
 * @return {Buffer}
 */
function encode(data) {
    return new buffer_1.Buffer(_encode(data));
}
exports.encode = encode;
/*
Decoding
*/
var Delimeters;
(function (Delimeters) {
    Delimeters[Delimeters["i"] = 105] = "i";
    Delimeters[Delimeters["e"] = 101] = "e";
    Delimeters[Delimeters["l"] = 108] = "l";
    Delimeters[Delimeters["d"] = 100] = "d";
    Delimeters[Delimeters["colon"] = 58] = "colon";
})(Delimeters || (Delimeters = {}));
function decodeString(data, encoding) {
    var firstColonIndex = null;
    for (var i = 0; i < data.length; i++) {
        if (data[i] === Delimeters.colon) {
            firstColonIndex = i;
            break;
        }
    }
    if (!firstColonIndex) {
        throw new Error("Error while decoding " + data.toString(encoding) + ". Missing colon.");
    }
    var expectedLength = parseInt(data.slice(0, firstColonIndex).toString());
    var value = data.slice(firstColonIndex + 1, firstColonIndex + 1 + expectedLength);
    return { value: value.toString(encoding), rest: data.slice(firstColonIndex + 1 + expectedLength) };
}
function decodeInteger(data) {
    var endIndex = null; // The first 'e' symbol index
    for (var i = 1; i < data.length; i++) {
        if (data[i] === Delimeters.e) {
            endIndex = i;
            break;
        }
    }
    if (!endIndex) {
        throw new Error("Invalid data. 'e' symbol expected");
    }
    return {
        value: parseInt(data.slice(1, endIndex).toString()),
        rest: data.slice(endIndex + 1)
    };
}
function decodeList(data, encoding) {
    // TODO: Check for 'e' symbol in the end.
    var result = [];
    var rest = data.slice(1); // l...
    var value = null;
    while (true) {
        var firstByte = rest[0];
        if (firstByte === Delimeters.i) {
            (_a = decodeInteger(rest), value = _a.value, rest = _a.rest, _a);
            result.push(value);
            continue;
        }
        if (encodesDigit(firstByte)) {
            (_b = decodeString(rest, encoding), value = _b.value, rest = _b.rest, _b);
            result.push(value);
            continue;
        }
        if (firstByte === Delimeters.l) {
            (_c = decodeList(rest, encoding), value = _c.value, rest = _c.rest, _c);
            result.push(value);
            continue;
        }
        if (firstByte === Delimeters.d) {
            (_d = decodeDict(rest, encoding), value = _d.value, rest = _d.rest, _d);
            result.push(value);
            continue;
        }
        if (firstByte === Delimeters.e) {
            rest = rest.slice(1);
            break;
        }
        throw new Error("Expected d, i, l or digit, got " + rest.toString(encoding));
    }
    return { value: result, rest: rest };
    var _a, _b, _c, _d;
}
function decodeDict(data, encoding) {
    var result = {};
    var rest = data.slice(1); // d...
    var value = null;
    var key;
    while (rest.length !== 0) {
        var firstByte = rest[0];
        if (firstByte === Delimeters.e) {
            rest = rest.slice(1);
            break;
        }
        (_a = decodeString(rest, encoding), key = _a.value, rest = _a.rest, _a);
        (_b = _decode(rest, encoding), value = _b.value, rest = _b.rest, _b);
        result[key] = value;
    }
    return { value: result, rest: rest };
    var _a, _b;
}
function encodesDigit(x) {
    return (x >= 0x30) && (x <= 0x39);
}
function _decode(data, encoding) {
    var value = null;
    var rest = null;
    var firstByte = data[0];
    if (firstByte === Delimeters.i) {
        return decodeInteger(data);
    }
    else if (encodesDigit(firstByte)) {
        return decodeString(data, encoding);
    }
    else if (firstByte === Delimeters.l) {
        return decodeList(data, encoding);
    }
    else if (firstByte === Delimeters.d) {
        return decodeDict(data, encoding);
    }
    else {
        throw new Error("Expected d, i, l or digit, got \"" + data.toString() + "\"");
    }
}
/**
 * Decodes bencoded data.
 *
 * @param {Buffer} data
 * @param {String} encoding (optional)
 * @return {Object|Array|String|Number}
 */
function decode(data, encoding) {
    var value;
    var rest;
    (_a = _decode(data, encoding), value = _a.value, rest = _a.rest, _a);
    if (rest.length) {
        throw new Error("Unexpected continuation: \"" + rest.toString() + "\"");
    }
    return value;
    var _a;
}
exports.decode = decode;
